examples/python/cancellation/README.md

```diff
@@ -0,0 +1,179 @@
+
+A client may cancel an RPC for several reasons. Perhaps the data it requested
+has been made irrelevant. Perhaps you, as the client, want to be a good citizen
+of the server and are conserving compute resources.
+
+##### Cancelling a Server-Side Unary RPC from the Client
+
+The default RPC methods on a stub will simply return the result of an RPC.
+
+```python
+>>> stub = hash_name_pb2_grpc.HashFinderStub(channel)
+>>> stub.Find(hash_name_pb2.HashNameRequest(desired_name=name))
+<hash_name_pb2.HashNameResponse object at 0x7fe2eb8ce2d0>
+```
+
+But you may use the `future()` method to receive an instance of `grpc.Future`.
+This interface allows you to wait on a response with a timeout, add a callback
+to be executed when the RPC completes, or to cancel the RPC before it has
+completed.
+
+In the example, we use this interface to cancel our in-progress RPC when the
+user interrupts the process with ctrl-c.
+
+```python
+stub = hash_name_pb2_grpc.HashFinderStub(channel)
+future = stub.Find.future(hash_name_pb2.HashNameRequest(desired_name=name))
+def cancel_request(unused_signum, unused_frame):
+    future.cancel()
+signal.signal(signal.SIGINT, cancel_request)
+```
+
+It's also important that you not block indefinitely on the RPC. Otherwise, the
+signal handler will never have a chance to run.
+
+```python
+while True:
+    try:
+        result = future.result(timeout=_TIMEOUT_SECONDS)
+    except grpc.FutureTimeoutError:
+        continue
+    except grpc.FutureCancelledError:
+        break
+    print("Got response: \n{}".format(result))
+    break
+```
+
+Here, we repeatedly block on a result for up to `_TIMEOUT_SECONDS`. Doing so
+gives the signal handlers a chance to run. In the case that our timeout
+was reached, we simply continue on in the loop. In the case that the RPC was
+cancelled (by our user's ctrl+c), we break out of the loop cleanly. Finally, if
+we received the result of the RPC, we print it out for the user and exit the
+loop.
+
+
+##### Cancelling a Server-Side Streaming RPC from the Client
+
+Cancelling a Server-side streaming RPC is even simpler from the perspective of
+the gRPC API. The default stub method is already an instance of `grpc.Future`,
+so the methods outlined above still apply. It is also a generator, so we may
+iterate over it to yield the results of our RPC.
+
+```python
+stub = hash_name_pb2_grpc.HashFinderStub(channel)
+result_generator = stub.FindRange(hash_name_pb2.HashNameRequest(desired_name=name))
+def cancel_request(unused_signum, unused_frame):
+    result_generator.cancel()
+signal.signal(signal.SIGINT, cancel_request)
+```
+
+However, the streaming case is complicated by the fact that there is no way to
+propagate a timeout to Python generators. As a result, simply iterating over the
+results of the RPC can block indefinitely and the signal handler may never run.
+Instead, we iterate over the generator on another thread and retrieve the
+results on the main thread with a synchronized `Queue`.
```

Comment:
Same as above, this is a workaround for our implementation defect. It's not that ideal to recommend it to our users...

Reply:
Agree it would be better to fix the bug. But until then, a workaround is the next best thing. This is an example. It can always be changed once our library supports a simpler workflow.